sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMemoryLeaseManager.scala

```diff
@@ -0,0 +1,406 @@
+ * If there isn't enough memory to even do that, then we can fall back to making that entire input
+ * batch spillable and calling the blocking API to get everything that we think we need to complete
+ * the operation.
+ */
+object GpuMemoryLeaseManager extends Logging {
+  // DO NOT ACCESS DIRECTLY!  Use `getInstance` instead.
+  @volatile private var instance: GpuMemoryLeaseManager = _
+
+  private def getInstance: GpuMemoryLeaseManager = {
+    if (instance == null) {
+      GpuMemoryLeaseManager.synchronized {
+        // The instance is trying to be used before it is initialized.
+        // Since we don't have access to a configuration object here,
+        // default to only one task per GPU behavior.
+        if (instance == null) {
+          logWarning("GpuMemoryLeaseManager was not properly initialized before use")
+          initialize(0)
+        }
+      }
+    }
+    instance
+  }
+
+  /**
+   * Initializes the GPU memory lease manager
+   * @param reservedMemory the amount of memory that is saved for other operations, right now
+   *                       this really is just for shuffle, but could expand to more in the future.
+   */
+  def initialize(reservedMemory: Long): Unit = synchronized {
+    if (instance != null) {
+      throw new IllegalStateException("already initialized")
+    }
+    val pool = if (Rmm.isPoolingEnabled) {
+      Rmm.getPoolSize
+    } else {
+      Cuda.memGetInfo().free
+    }
+    instance = new GpuMemoryLeaseManager(pool - reservedMemory)
+  }
+
+  /**
+   * Initialization intended to only be used for testing
+   * @param poolSize the size of the pool
+   */
+  def initializeForTest(poolSize: Long): Unit = synchronized {
+    if (instance != null) {
+      shutdown()
+    }
+    instance = new GpuMemoryLeaseManager(poolSize)
+  }
+
+  /**
+   * Request a lease to access more GPU memory. The thread may block until enough memory can
+   * be made available. All memory for this task must be spillable before this is called,
+   * because if the task blocks that memory can be handed out to other tasks to make progress.
+   *
+   * @param tc the task that is making the request.
+   * @param optionalAmount the amount of memory that is desired. If this amount of memory is not
+   *                       available because the GPU does not have that much memory the request may
+   *                       be lowered. This is on top of any required amount of memory.
+   * @param requiredAmount the amount of memory that is required. This is on top of any optional
+   *                       amount of memory that is requested. If the GPU does not have enough
+   *                       memory to fulfill this request an exception will be thrown.
+   * @return a MemoryLease that allows the task to allocate more memory.
+   */
+  def requestLease(tc: TaskContext, optionalAmount: Long, requiredAmount: Long = 0): MemoryLease =
```

Comment:
naming ideas: use maximumAmount instead of optionalAmount and minimumAmount instead of requiredAmount?  that seems clearer than optional/required.  and if you specify both maximumAmount and minimumAmount, then maximumAmount would be greater than minimumAmount.  seems confusing where existing parameters as independent (e.g. "this is on top of any required amount of memory").

Reply:
I think this api suggestion makes it clearer as well and ports well for the non-blocking lease.